src/coreclr/jit/promotion.cpp

```diff
@@ -378,52 +385,78 @@ class LocalUses
         countReadBacksWtd += countOverlappedRetbufsWtd;
-        countReadBacksWtd += countOverlappedAssignmentDestinationWtd;
 
-        // A read back puts the value from stack back to (hopefully) register. We cost it at 5 bytes.
-        costWith += countReadBacksWtd * 5;
+        // The same if the struct was assigned from a call, since we don't
+        // currently have any "forwarding" optimization for this case.
+        countReadBacksWtd += countOverlappedAssignedFromCallWtd;
+
+        // A readback turns into a stack load that we costed at 3 above.
+        costWith += countReadBacksWtd * 3;
 
         // Write backs with TYP_REFs when the base local is an implicit byref
-        // involves checked write barriers, so they are very expensive.
+        // involves checked write barriers, so they are very expensive. We cost that at 10 cycles.
         // TODO-CQ: This should be adjusted once we type implicit byrefs as TYP_I_IMPL.
-        weight_t writeBackCost = comp->lvaIsImplicitByRefLocal(lclNum) && (access.AccessType == TYP_REF) ? 15 : 5;
-        weight_t countWriteBacksWtd =
-            countOverlappedCallsWtd + countOverlappedReturnsWtd + countOverlappedAssignmentSourceWtd;
+        // Otherwise we cost it like a store to stack at 3 cycles.
+        weight_t writeBackCost = comp->lvaIsImplicitByRefLocal(lclNum) && (access.AccessType == TYP_REF) ? 10 : 3;
+
+        // We write back before an overlapping struct use passed as an arg.
+        // TODO-CQ: A store-forwarding optimization in lowering could get rid
+        // of these copies; however, it requires lowering to be able to prove
+        // that not writing the fields into the struct local is ok.
+        weight_t countWriteBacksWtd = countOverlappedCallArgWtd;
         costWith += countWriteBacksWtd * writeBackCost;
 
+        // Overlapping assignments are decomposable so we don't cost them as
+        // being more expensive than their unpromoted counterparts (i.e. we
+        // don't consider them at all). However, we should do something more
+        // clever here, since:
+        // * We may still end up writing the full remainder as part of the
+        //   decomposed assignment, in which case all the field writes are just
+        //   added code size/perf cost.
+        // * Even if we don't, decomposing a single struct write into many
+        //   field writes is not necessarily profitable (e.g. 16 byte field
+        //   stores vs 1 XMM load/store).
+        //
+        // TODO-CQ: This ends up being a combinatorial optimization problem. We
+        // need to take a more "whole-struct" view here and look at sets of
+        // fields we are promoting together, evaluating all of them at once in
+        // comparison with the covering struct uses. This will also allow us to
+        // give a bonus to promoting remainders that may not have scalar uses
+        // but will allow fully decomposing assignments away.
```

Comment:
Sounds ambitious, but I like the idea of costing things as sets.

Reply:
Yeah, hopefully I can approximate this in a simple/cheap way. But I think I will approach this by writing down an integer linear programming form of the problem and see if it gives me any insights.